### PR TITLE
#477 query to native sql fix

### PR DESCRIPTION
**Pull request by cserhatit**

### CODE REVIEW

File: coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/JpaUtil.java [hunk 1]
```diff
@@ -37,80 +45,123 @@
...
+     *
+     * @param criteriaQuery
+     *            created query
+     * @return native SQL of query
+     * @throws BaseException
+     *             on error
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public static String toNativeSQL(Query criteriaQuery) throws BaseException {
+        QueryImplementor query = criteriaQuery.unwrap(QueryImplementor.class);
```
Comment (rombow) [inline coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/JpaUtil.java:118]: npe fix plz

File: coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/JpaUtil.java [hunk 2]
```diff
@@ -37,80 +45,123 @@
...
+        if (query instanceof SqmInterpretationsKey.InterpretationsKeySource && query instanceof QuerySqmImpl) {
+            QueryInterpretationCache.Key cacheKey = SqmInterpretationsKey
+                    .createInterpretationsKey((SqmInterpretationsKey.InterpretationsKeySource) query);
+            QuerySqmImpl<?> querySqm = (QuerySqmImpl<?>) query;
+            Supplier buildSelectQueryPlan = () -> {
+                try {
+                    return ReflectionUtils.invokeMethod(querySqm, "buildSelectQueryPlan", ConcreteSqmSelectQueryPlan.class);
+                } catch (BaseException e) {
+                    Logger.getLogger(JpaUtil.class).warn("Exception on calling buildSelectQueryPlan()!", e);
+                    return null;
```
Comment (rombow) [inline coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/JpaUtil.java:128]: why doesn't it throw the exception but return null instead?

File: coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/ReflectionUtils.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+     * @param targetClass
+     *            the provided Java {@link Class} the field belongs to
+     * @param fieldName
+     *            the {@link Field} name
+     * @return the {@link Field} matching the given name
+     * @throws BaseException
+     *             on error
+     */
+    public static Field getField(Class<?> targetClass, String fieldName) throws BaseException {
+        Field field = null;
```
Comment (rombow) [inline coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/ReflectionUtils.java:56]: npe check plz to all public method

**Review by rombow — CHANGES_REQUESTED** [review]

File: coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/ReflectionUtils.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+import hu.icellmobilsoft.coffee.dto.exception.TechnicalException;
+import hu.icellmobilsoft.coffee.dto.exception.enums.CoffeeFaultType;
+
+/**
+ * <code>ReflectionUtils</code> - Reflection utilities holder from hypersistence-utils-60
+ *
+ * @author tamas.cserhati
+ * @since 2.1.0
+ */
+@Vetoed
```
Comment (speter555) [inline coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/ReflectionUtils.java:38]: @Vetoed annotation was no longer required.

File: coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/JpaUtil.java (hunk 2 above)
Comment (cserhatit) [reply coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/JpaUtil.java:131]: Supplier<T> cannot throw checked exception and next step needs the Supplier<T> interface as parameter (so i couldnt use own supplier with baseexception )

**Review by rombow — APPROVED** [review]

**Review by czenczl — APPROVED** [review]

File: coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/JpaUtil.java [hunk 5]
```diff
@@ -64,53 +91,80 @@ public static String toNativeSQLNoEx(EntityManager em, Query query) {
...
+                            .getFactory()
+                            .getQueryEngine()
+                            .getInterpretationCache()
+                            .resolveSelectQueryPlan(cacheKey, buildSelectQueryPlan)
+                    : (SelectQueryPlan<?>) buildSelectQueryPlan.get();
+            if (plan instanceof ConcreteSqmSelectQueryPlan) {
+                ConcreteSqmSelectQueryPlan<?> selectQueryPlan = (ConcreteSqmSelectQueryPlan<?>) plan;
+                Object cacheableSqmInterpretation = ReflectionUtils.getFieldValueOrNull(selectQueryPlan, "cacheableSqmInterpretation", Object.class);
+                if (cacheableSqmInterpretation == null) {
+                    DomainQueryExecutionContext domainQueryExecutionContext = DomainQueryExecutionContext.class.cast(querySqm);
```
[the remark this answers is not in the document]
Comment (cserhatit) [reply coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/util/JpaUtil.java:145]: it is safe to cast from: QuerySqmImpl implements DomainQueryExecutionContext

**Review by petrenyi-mark — APPROVED** [review]

**Pull request merged**